### PR TITLE
766 implementation of ode sir model

### LINKED ISSUE #766: implementation of ODE SIR model
### Feature description

c++ implementation of SIR model with simple example and python bindings 

### Additional context

_No response_

### Checklist

- [x] Attached labels, especially loc:: or model:: labels.
- [x] Linked to project.

### PR DESCRIPTION
**Pull request by alexgit0**
# Changes and Information

- Implementation of SIR model and corresponding python bindings 

## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/DLR-SC/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/DLR-SC/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added

### Checks by code reviewer(s)

- [x] Corresponding issue(s) is/are linked and addressed
- [x] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [x] Appropriate **unit tests** have been added, CI passes and code coverage is acceptable (did not decrease)
- [x] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)

closes #766

### CODE REVIEW

File: cpp/models/ode_sir/parameters.h [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    static std::string name()
+    {
+        return "ContactPatterns";
+    }
+};
+
+using ParametersBase = ParameterSet<TransmissionProbabilityOnContact, TimeInfected, ContactPatterns>;
+
+/**
+ * @brief Parameters of an age-resolved SECIR/SECIHURD model.
```
Comment (MaxBetzDLR) [inline cpp/models/ode_sir/parameters.h:68]: Wrong model in brief

File: cpp/models/ode_sir/parameters.h [hunk 2]
```diff
@@ -0,0 +1,162 @@
+
```
Comment (MaxBetzDLR) [inline cpp/models/ode_sir/parameters.h:1]: Copyright header missing

File: cpp/models/ode_sir/model.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
+
```
Comment (MaxBetzDLR) [inline cpp/models/ode_sir/model.h:1]: Copyright header

File: cpp/models/ode_sir/model.cpp [hunk 4]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (MaxBetzDLR) [inline cpp/models/ode_sir/model.cpp:1]: Copyright header

File: cpp/models/ode_sir/infection_state.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (MaxBetzDLR) [inline cpp/models/ode_sir/infection_state.h:1]: Copyright header

File: cpp/examples/ode_sir.cpp [hunk 6]
```diff
@@ -0,0 +1,68 @@
+
```
Comment (MaxBetzDLR) [inline cpp/examples/ode_sir.cpp:1]: copyright header missing. You can use the one from ode_seir/model.cpp

File: cpp/tests/data/sir-js-compare.csv
Comment (MaxBetzDLR) [inline cpp/tests/data/sir-js-compare.csv:1]: Prefix ode should be added, as in the file name ide-seir-compare.csv

File: pycode/examples/simulation/osir_simple.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
+#############################################################################
+# Copyright (C) 2020-2021 German Aerospace Center (DLR-SC)
```
Comment (MaxBetzDLR) [inline pycode/examples/simulation/osir_simple.py:2]:
```suggestion
# Copyright (C) 2020-2023 German Aerospace Center (DLR-SC)
```

File: pycode/memilio-simulation/memilio/simulation/osir.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
+
```
Comment (MaxBetzDLR) [inline pycode/memilio-simulation/memilio/simulation/osir.cpp:1]: Copyright header

File: pycode/memilio-simulation/memilio/simulation/osir.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
+#############################################################################
+# Copyright (C) 2020-2021 German Aerospace Center (DLR-SC)
```
Comment (MaxBetzDLR) [inline pycode/memilio-simulation/memilio/simulation/osir.py:2]:
```suggestion
# Copyright (C) 2020-2023 German Aerospace Center (DLR-SC)
```

File: pycode/memilio-simulation/memilio/simulation_test/test_osir.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
+#############################################################################
+# Copyright (C) 2020-2021 German Aerospace Center (DLR-SC)
```
Comment (MaxBetzDLR) [inline pycode/memilio-simulation/memilio/simulation_test/test_osir.py:2]:
```suggestion
# Copyright (C) 2020-2023 German Aerospace Center (DLR-SC)
```

**Review by MaxBetzDLR — CHANGES_REQUESTED** [review]

File: cpp/tests/data/sir-js-compare.csv
Comment (alexgit0) [reply cpp/tests/data/sir-js-compare.csv:1]: prefix will be added, however there is no prefix for seir-js-compare.csv either

Comment (MaxBetzDLR) [reply cpp/tests/data/sir-js-compare.csv:1]: Your right, if possible change it aswell

File: cpp/examples/ode_sir.cpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    mio::osir::Model model;
+
+    model.populations[{mio::Index<mio::osir::InfectionState>(mio::osir::InfectionState::Infected)}]  = 1000;
+    model.populations[{mio::Index<mio::osir::InfectionState>(mio::osir::InfectionState::Removed)}] = 1000;
+    model.populations[{mio::Index<mio::osir::InfectionState>(mio::osir::InfectionState::Susceptible)}] =
+        total_population -
+        model.populations[{mio::Index<mio::osir::InfectionState>(mio::osir::InfectionState::Infected)}] -
+        model.populations[{mio::Index<mio::osir::InfectionState>(mio::osir::InfectionState::Removed)}];
+    // suscetible now set with every other update
+    // params.nb_sus_t0   = params.nb_total_t0 - params.nb_exp_t0 - params.nb_inf_t0 - params.nb_rec_t0;
```
Comment (mknaranja) [inline cpp/examples/ode_sir.cpp:48]: exposed does not exist here

File: cpp/examples/ode_sir.cpp [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    model.parameters.set<mio::osir::TimeInfected>(2);
+    model.parameters.set<mio::osir::TransmissionProbabilityOnContact>(1);
+    model.parameters.get<mio::osir::ContactPatterns>().get_baseline()(0, 0) = 2.7;
+    model.parameters.get<mio::osir::ContactPatterns>().add_damping(0.6, mio::SimulationTime(12.5));
+
+    auto integrator                          = std::make_shared<mio::EulerIntegratorCore>();
+
+    model.check_constraints();
+    // print_seir_params(model);
```
Comment (mknaranja) [inline cpp/examples/ode_sir.cpp:57]: pls remove development artifacts

File: cpp/models/ode_sir/infection_state.h
[the remark this answers is not in the document]
Comment (mknaranja) [reply cpp/models/ode_sir/infection_state.h:1]: Please use `Recovered` as in SEIR.

File: cpp/models/ode_sir/model.h [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    void get_derivatives(Eigen::Ref<const Eigen::VectorXd> pop, Eigen::Ref<const Eigen::VectorXd> y, double t,
+                         Eigen::Ref<Eigen::VectorXd> dydt) const override
+    {
+        auto& params     = this->parameters;
+        double coeffStoE = params.get<ContactPatterns>().get_matrix_at(t)(0, 0) *
+                           params.get<TransmissionProbabilityOnContact>() / populations.get_total();
+
+        dydt[(size_t)InfectionState::Susceptible] = 
+            -coeffStoE * y[(size_t)InfectionState::Susceptible] * pop[(size_t)InfectionState::Infected];
```
Comment (mknaranja) [inline cpp/models/ode_sir/model.h:57]: The variable should be `coeffStoI` now.

Please think of adapting related comments and variable names too when changing the model.

File: cpp/tests/data/ode-sir-js-compare.csv [hunk 14]
```diff
@@ -0,0 +1,501 @@
+ # t S I R
```
Comment (mknaranja) [inline cpp/tests/data/ode-sir-js-compare.csv:1]: Although it is not 100kb, 40kb is still large for a file like this. You probably adapted well from the old file but we should stick to smaller files as the repo grows. I think comparing the first 20 steps is already sufficient to compare the results to some higher accuracy. If there is an imprecision of X after 100 iterations, there is an imprecision of Y after 10 iterations.

File: cpp/models/ode_sir/README.md [hunk 15]
```diff
@@ -0,0 +1,20 @@
+
+# ODE SIR compartment model
+
+This model is a very simple ODE model with only three compartments and few parameters, mostly for demonstration of the MEmilio framework:
+- Susceptible, may become infected at any time
+- Infected, will be recovered after some time
+- Recovered, recovered from infectious process (dead or recovered).
```
Comment (MaxBetzDLR) [inline cpp/models/ode_sir/README.md:7]:
```suggestion
- Recovered, recovered from infectious process (dead or recovered)

```

**Review by HenrZu — APPROVED** [review]

File: cpp/models/ode_sir/infection_state.h [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#ifndef SIR_INFECTIONSTATE_H
+#define SIR_INFECTIONSTATE_H
```
Comment (mknaranja) [inline cpp/models/ode_sir/infection_state.h:22]:
```suggestion
#ifndef ODESIR_INFECTIONSTATE_H
#define ODESIR_INFECTIONSTATE_H
```

File: cpp/models/ode_sir/infection_state.h [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    Susceptible,
+    Infected,
+    Recovered,
+    Count
+};
+
+} // namespace osir
+} // namespace mio
+
+#endif // SIR_INFECTIONSTATE_H
```
Comment (mknaranja) [inline cpp/models/ode_sir/infection_state.h:44]:
```suggestion
#endif // ODESIR_INFECTIONSTATE_H
```

**Review by MaxBetzDLR — APPROVED** [review]

File: cpp/models/ode_sir/model.h [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#ifndef SIR_MODEL_H
+#define SIR_MODEL_H
```
Comment (mknaranja) [inline cpp/models/ode_sir/model.h:22]:
```suggestion
#ifndef ODESIR_MODEL_H
#define ODESIR_MODEL_H
```

File: cpp/models/ode_sir/model.h [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+            (1.0 / params.get<TimeInfected>()) * y[(size_t)InfectionState::Infected];
+        dydt[(size_t)InfectionState::Recovered] =
+            (1.0 / params.get<TimeInfected>()) * y[(size_t)InfectionState::Infected];
+    }
+};
+
+} // namespace osir
+} // namespace mio
+
+#endif // SIR_MODEL_H
```
Comment (mknaranja) [inline cpp/models/ode_sir/model.h:69]:
```suggestion
#endif // ODESIR_MODEL_H
```

**Review by mknaranja — CHANGES_REQUESTED** [review]

**Pull request merged**